### ISSUE TITLE
sasl not working

### ISSUE THREAD

**Issue by kohlrak** [cnaude/PurpleIRC-spigot#61]
I use TCP dump and tried to find exactly why I kept not logging me in and it turns out that sasl is not sending The exact information being put in. Copying and pasting the password to ident password and it works. My server correctly forwards it for other irc clients.my suspicion is that one of the characters holds significance in yml, seeing as it refuses to keep the single quotes around it was the result of a keyboard smash.

**Comment by cnaude** [comment]
Yaml can be a little tricky with string formatting. I would recommend putting single quotes around the password in the yml file. This should be able to handle most use cases. If your password has a single quote in it then you can use double quotes around the password instead. 
```yaml
ident-password: 'password'
```

**Comment by kohlrak** [comment]
> Yaml can be a little tricky with string formatting. I would recommend putting single quotes around the password in the yml file. This should be able to handle most use cases. If your password has a single quote in it then you can use double quotes around the password instead.
> 
> ```yaml
> ident-password: 'password'
> ```

I tried doing that. I noticed that every time I tried modifying it and reloading within Minecraft it would erase the quotes. Or is it one of those we can't modify and reload when running?

**Comment by cnaude** [comment]
I was thinking about how to resolve this issue and I wonder if using a file reference to a password file would help. The way that the yaml library updates the yaml file can sometimes cause issues like this, especially when there are a bunch of special characters that need to be escaped.

**Comment by kohlrak** [comment]
I'm not sure if that's the case. I forget what it was, but i looked further
into it, and noticed it does indeed remove the quotes, that was not user
error like i suspected. Are you using sasl?

**Comment by cnaude** [comment]
I don't currently use sasl. The yaml library is likely just not handling one of the special characters properly. I can create a version of the plugin it's able to load the sasl password from another file instead of from the yaml. Would you be willing to test that if I create it?

**Comment by kohlrak** [comment]
Yeah, I can do that! There may be a delay in doing so, though. As you might
notice from my response rate. I plan on going live this weekend so if you
get me a build by then you'll have a test by then.

2024年5月7日(火) 16:25 Chris ***@***.***>:

**Comment by cnaude** [comment]
https://jenkins.chrisnaude.com/jenkins/job/PurpleIRC-spigot/379/

This new version has a new option in the bot yaml:
```yaml
sasl-password-file: password.txt
```

**Comment by kohlrak** [comment]
All right I'll test you soon as I get home I'm at work now there was a
scheduling shift change this weekend everything went live but the my
world's plugin that I rely on just updated as well so I had some other
things to test at the same time.

2024年5月12日(日) 17:36 Chris ***@***.***>:

**Comment by kohlrak** [comment]
Got some time here at work, i couldn't get it to load the password file. I
even tried an absolute path. I chowned and chgrped everything too


2024年5月12日(日) 21:55 kohlrak ***@***.***>:

**Comment by cnaude** [comment]
Oh that's really strange. Did it throw any errors? When I tested it I had
the password file in the same location as the bot yml file.

**Comment by kohlrak** [comment]
nvm, i was in a hurry and testing other mods at the same time: i misread
teh error message about the Craft Scheduler Thread. I don't have enough
time to try at the moment so i'll try again when i get home and can look at
it. It finally hit me after i set the sasl to false. I leave work in 2
hours.

2024年5月13日(月) 1:03 Chris ***@***.***>:

> Oh that's really strange. Did it throw any errors? When I tested it I had
> the password file in the same location as the bot yml file.
>

**Comment by kohlrak** [comment]
Ok, i found the problem, sort of. I even removed special characters from
the password, i've used the file, i've set the thing to plaintext using
sasl and i finally used tcpdump to see what's going on: it's not sending
the sasl data. It's sending something, but whatever it is it's not the
correct data. I don't know how to read the packet info, but it looks to me
like it's just sending null data:

I looked deeper into the issue, and apparently it's a known issue with
pircbotx. There seems to be a conflict between the creators of pircbotx and
inspIRCd. It's part of a bigger discussion and looks like an unnecessary
conflict of finger pointing. Servers like inspIRCd (which i'm using, as
well as a few other very well known networks like Freenode (i assume Libera
as well since that spawned form Freenode, but i can't confirm)) are using
"AUTHENTICATE :+" where as the official IRCv3 spec makes it unclear whether
or not the ":" is optional and ignorable. So what's happening is because
the server isn't sending "AUTHENTICATE +" without the colon, pircbotx is
just sending null data to keep the connection alive while it waits for that
message which never comes. The github for the spec has implicitly taken a
side on the issue, but it hasn't explicitly stated how it should be
handled. Basically, it's needless drama.

It would appear this issue is outside of the scope of PurpleIRC. But i
learned something new and have to change one of my C libraries when i get
more time. For the time being i'm going to have to use nickserv
authentication, even if it's less secure, because the responsibility has
been clarified back in 2021 and the fact the issue hasn't been solved
suggests it's not likely to see the light of day any time soon.

**Comment by cnaude** [comment]
I've been maintaining my own version of pircbotx so I might be able to fix
that.

**Comment by kohlrak** [comment]
Well, in tht case, if you haven't modified the file, the issue lies on line
67 of src/main/java/org/pircbotx/cap/SASLCapHandler.java.

Since i'm so out of the loop with java since i haven't used it since 2008
or 2009, i asked chatgpt. According to it: The String.replace() method in
Java does not throw an exception if it fails to find the specified
character or substring to replace. If the character or substring is not
found, it simply returns the original string unchanged. So, in this case,
if there are no semicolons in the string, it will return the original
string as it is.

I asked it about colons, not semicolons, but that's why we don't ask it to
actually code for us.

**Comment by cnaude** [comment]
I've created test version with an updated pircbotx: https://www.dropbox.com/scl/fi/hva9v4wo3wd5ch9l7y93b/PurpleIRC.jar?rlkey=f7tvak56i4z0hei6u1e78ho6o&dl=0

**Comment by kohlrak** [comment]
That did it. I tested both the file method and actually using it directly
from the config file. And it's working in both scenarios. Thank you.

**Comment by cnaude** [comment]
Well [PircBotX](https://jenkins.chrisnaude.com/jenkins/job/PircBotX/) now compiles on JDK21. :D
[PurpleIRC-1.0-381](https://www.spigotmc.org/resources/purpleirc.2836/) has been uploaded and contains the fix.
